### PR TITLE
ADR: Configuration of import aliases in Javascript applications

### PR DESCRIPTION
**Pull request by yordis**

### CODE REVIEW

File: adrs/2068541249/README.md [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+part of our codebase?
+
+Although it may be simply to answer the question by scanning the directories and
+files under `./src`.
+
+In other cases like `import { useMyQuery } from 'react-query';` where you
+may have `react-query` as a dependency, and you used `./src/react-query`
+directory to store your React Query related things as an example, you need to
+scan the files to make sure that the imports don't belong to a dependency
+package.
```
Comment (bionikspoon) [inline adrs/2068541249/README.md:81]: Great point!  In other words, you could inadvertently override an external module locally.

Comment (bionikspoon) [reply adrs/2068541249/README.md:81]: I don't know if this story is relevant to Typescript. Still, In the past, I've experienced a scenario where external dependencies had a common name like `http-client` and that 3rd party dependency was resolving to my local `src/http-client/index`.  This is the nightmare scenario.

File: adrs/2068541249/README.md [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+    ```jsonc
+    {
+      "compilerOptions": {
+        "baseUrl": ".", // The baseUrl is at the root of the project
+        "paths": {
+          "@/*": ["./src/*"]
+        }
+      }
+    }
+    ```
```
Comment (bionikspoon) [inline adrs/2068541249/README.md:149]: Would it be out of scope to include how to set up [`eslint-plugin-import`](https://github.com/benmosher/eslint-plugin-import#resolvers) to respect this setup?

Comment (yordis) [reply adrs/2068541249/README.md:149]: We could add some docs about it for sure

**Review by bionikspoon — APPROVED** [review]
I like to throw my best counterargument at these ADRs.  I can't think of one for this ADR. This ADR is rock-solid advice.

File: adrs/2068541249/README.md (hunk 1 above)
Comment (yordis) [reply adrs/2068541249/README.md:81]: Yes, sir, the error-prune is really high with such configuration.

File: adrs/2068541249/README.md (hunk 2 above)
Comment (yordis) [reply adrs/2068541249/README.md:149]: Actually, I am gonna keep ESLint out of the scope of the ADR.

File: adrs/2068541249/README.md [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+* You **must** never change `baseUrl` from the root of the project.
+* You **must** use the following configuration to resolve files relative to the
+  root of the project:
+
+    ```jsonc
+    {
+      "compilerOptions": {
+        "baseUrl": ".", // The baseUrl is at the root of the project
+        "paths": {
+          "@/*": ["./src/*"]
```
Comment (tleef) [inline adrs/2068541249/README.md:145]: I actually ran into an issue with this recently. I can't remember it off the top of my head. I'll see if I can dig it up.

**Review by mboperator — APPROVED** [review]

File: adrs/2068541249/README.md (hunk 3 above)
Comment (tleef) [reply adrs/2068541249/README.md:145]: https://github.com/facebook/create-react-app/issues/9406

**Comment by yordis** [comment]
It seems that CRA didn't take into consideration that the following package could exists https://www.npmjs.com/package/src since https://github.com/facebook/create-react-app/blob/8bf050aa7c16078fed5e51ac8388d6100c29e105/packages/react-scripts/config/modules.js#L76

File: adrs/2068541249/README.md (hunk 3 above)
Comment (yordis) [reply adrs/2068541249/README.md]: Following the issue, I changed the ADR to be more generic and agnostic of `{ts,js}config.json` but to be more focused on the issue and guidelines about it.

**Comment by yordis** [comment]
Moved the ADR to Reviewing

**Comment by tleef** [comment]
I found a decent workaround for CRA which I think satisfies the spirit of this ADR.

The problem is CRA doesn't allow you to define path aliases. However, we are able to define a baseUrl in our `tsconfig.json` See [here](https://github.com/facebook/create-react-app/issues/9406)

We can use the baseUrl to achieve the goal of importing our application files like the following example
```
import { Header } from "@/components/Header";
```
CRA sets up the directory structure for you and requires that you have a directory named `src` with an `index.tsx` file and a `react-app-env.d.ts` file (assuming you are using the TypeScript template) It also comes with a predefined `tsconfig.json`

We can satisfy these requirements with minimal changes to the generated directory structure and `tsconfig.json` while still being able to import our app modules using the above style by setting our `"baseUrl": "./src"` in our `tsconfig.json` and creating a `@` directory inside of `src`

The resulting directory structure would be as follows
```text
tree . -L 2
├── tsconfig.json
├── src
│   ├── index.tsx
│   ├── react-app-env.d.ts
│   ├── @
│   │   ├── components/...
```
and our `tsconfig.json` will look like this (or similar)
```
{
  "compilerOptions": {
    "target": "es5",
    "lib": [
      "dom",
      "dom.iterable",
      "esnext"
    ],
    "allowJs": true,
    "skipLibCheck": true,
    "esModuleInterop": true,
    "allowSyntheticDefaultImports": true,
    "strict": true,
    "forceConsistentCasingInFileNames": true,
    "module": "esnext",
    "moduleResolution": "node",
    "resolveJsonModule": true,
    "isolatedModules": true,
    "noEmit": true,
    "jsx": "react",
    "baseUrl": "./src"
  },
  "include": [
    "src"
  ]
}
```

**Review by tleef — APPROVED** [review]

**Pull request merged**